### PR TITLE
Add BHVSpace and WaldSpaceMetric

### PR DESCRIPTION
**Pull request by luisfpereira**
This PR ports `WaldSpaceMetric` from `waldspaces` (developed by @JonasLueg) to `geomstats`.

### CODE REVIEW

File: geomstats/geometry/stratified/wald_space.py [hunk 1]
```diff
@@ -47,949 +15,973 @@ class :class:`spd.SPDMatrices`.
...
+)
+from geomstats.geometry.stratified.vectorization import (
+    _manipulate_output,
+    broadcast_lists,
+    vectorize_point,
+)
+from geomstats.numerics.optimizers import ScipyMinimize
+
+
+def _manipulate_input_with_array(arg):
```
Comment (ninamiolane) [inline geomstats/geometry/stratified/wald_space.py:44]: docstring?

File: geomstats/geometry/stratified/wald_space.py [hunk 2]
```diff
@@ -47,949 +15,973 @@ class :class:`spd.SPDMatrices`.
...
+
+    return arg, False
+
+
+def _manipulate_output_wald(out, to_list):
+    return _manipulate_output(out, to_list, manipulate_output_iterable=WaldCollection)
+
+
+def make_splits(n_labels):
+    """Generate all possible splits of a collection."""
```
Comment (ninamiolane) [inline geomstats/geometry/stratified/wald_space.py:61]: parameter description missing

File: geomstats/geometry/stratified/wald_space.py [hunk 3]
```diff
@@ -47,949 +15,973 @@ class :class:`spd.SPDMatrices`.
...
+    if n_labels <= 1:
+        raise ValueError("`n_labels` must be greater than 1.")
+    if n_labels == 2:
+        yield Split(part1=[0], part2=[1])
+    else:
```
Comment (ninamiolane) [inline geomstats/geometry/stratified/wald_space.py:66]: is else needed after yield?

File: geomstats/geometry/stratified/point_set.py [hunk 4]
```diff
@@ -1,91 +1,61 @@
...
+        is_equal : array-like, shape=[...]
+        """
 
-class Point(ABC):
-    r"""Class for points of a set."""
 
-    @abstractmethod
-    def __repr__(self):
-        """Produce a string with a verbal description of the point."""
+class PointCollection(ABC, list):
```
Comment (ninamiolane) [inline geomstats/geometry/stratified/point_set.py:30]: Explain what this is, explain that this is the analogous of a batch.
Maybe rename it PointBatch?

**Review by ninamiolane — CHANGES_REQUESTED** [review]
What is the plan to use the learning algorithms (that are compatible with distance-only structures) on these spaces?

NIT: not sure we need `_space.py` in file names

File: geomstats/geometry/stratified/wald_space.py (hunk 3 above)
Comment (luisfpereira) [reply geomstats/geometry/stratified/wald_space.py:66]: Yes, otherwise also runs code after else.

**Comment by luisfpereira** [comment]
@ninamiolane, I'm merging this one now as, while probably this still needs some refinement, it is in a much better state than we have available in main. Let's revisit it if you disagree with something.

**Pull request merged**